text/0517-io-os-reform.md

```diff
@@ -483,7 +483,72 @@ throughout IO, we can go on to explore the modules in detail.
 ### `std::env`
 [std::env]: #stdenv
 
-> To be added in a follow-up PR.
+Most of what's available in `std::os` today will move to `std::env`,
+and the signatures will be updated to follow this RFC's
+[Design principles] as follows.
+
+**Arguments**:
+
+* `args`: change to yield an iterator rather than vector if possible; in any case, it should produce an `OsStrBuf`.
+
+**Environment variables**:
+
+* `vars` (renamed from `env`): yields a vector of `(OsStrBuf, OsStrBuf)` pairs.
```

Comment:
It does depend on what platform you're running on, however. On Windows there's no extra allocation necessary as the OS hands you an already-allocated block of strings. On Unix we're basically doing the same thing as Windows, except we're the one implementing it (copying the `environ` into a locally owned allocation). So for something like `env::vars().collect::<Vec<_>>()` you would save an allocation on Unix if you instead returned a `Vec` but you would force an unnecessary extra one on Windows if you didn't need to collect to a vector.

The analogy is pretty much the same for `args` as well I believe. Windows will hand us an already-allocated block that we locally own but we're forced to create one on Unix ourselves. All that being said though, I do also think that consistency and API design is probably more important here than runtime performance as these probably aren't the source of slowdown for too many programs :)